### DISCUSSION TITLE
Memory offset of `ref bool` in `extern` method

### DISCUSSION THREAD

**Discussion by wegylexy** [dotnet/runtime#56328] [Q&A] [answered]
### Description
```cpp
extern "C"
{
    void MyNativeFunction(bool* flags);
}
```
```cs
[DllImport("MyNativeLib")]
static extern void MyNativeFunction(ref bool flags);

Span<bool> flags = stackalloc bool[24576];
// more code / stackalloc in between increases the offset
MyNativeFunction(ref MemoryMarshal.GetReference(flags));
// now, values of flags seen here are offset by arbitrary number of bytes
```
Adding `[MarshalAs(UnmanagedType.U1)]` does not help.

### Configuration

.NET 6 preview 6 Linux x64 docker image from `mcr.microsoft.com/dotnet/aspnet:6.0`

### Regression?

It works fine when casted to byte first, or with pointers.
```cs
[DllImport("MyNativeLib")]
static extern void MyNativeFunction(ref byte flags);

Span<bool> flags = stackalloc bool[24576];
MyNativeFunction(ref MemoryMarshal.GetReference(MemoryMarshal.Cast<bool, byte>(flags)));
```
```cs
[DllImport("MyNativeLib")]
static extern void MyNativeFunction(nint flags);

Span<bool> flags;
unsafe
{
    bool* p = stackalloc bool[24576];
    MyNativeFunction((nint)p);
    flags = new(p, 24576);
}
```

### Other information

Offsets that actually experienced were -96 bytes and -64 bytes.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by wegylexy** [comment]
I wonder if the marshaller uses a shadow copy elsewhere, passing the address of that memory location.
Is there a way to disable marshalling and pass `ref bool` by reference directly without casting to `IntPtr`/`nint` in `unsafe` and `bool*`, or to `ref byte` via `Span<byte>`?

**Comment by jkoritzinsky** [comment, marked as answer]
`bool` values are always marshalled when passed by value or by `ref`. The runtime does normalization on both managed->native and native->managed for `true` and `false`.

 You'll either need to use `bool*` (where we then assume that you know what you are doing and leave the code be), or pass as a `ref byte` using unsafe code.

In general, if you want to have the marshaller not make any copies into locals in a marshalling stub, do not use any non-blittable types.

**Reply by wegylexy** [reply]
Thanks. I didn't expect `bool` to be non-blittable.
Should I use unsafe code or `MemoryMarshal.Cast<bool, byte>()`?

**Reply by jkoritzinsky** [reply]
MemoryMarshal works. It’s technically unsafe code as well, so be careful for other usages (this one is usually fine).
